File: 2007. Find Original Array From Doubled Array.py
class Solution:
    def findOriginalArray(self, changed: List[int]) -> List[int]:
        changed.sort(reverse=True)
        start, end, demo = 0, 1, []
        count = Counter(changed)
        
        if changed.count(0) % 2 == 1:
            return []
        
        demo = [0 for i in range(changed.count(0) // 2)]
        count[0] -= changed.count(0)
        
        while start < len(changed) and end < len(changed):
            if count[changed[start]] == 0:
                start += 1
            elif changed[start] == 2 * changed[end] and count[changed[end]] != 0:
                demo.append(changed[end])
                count[changed[start]] -= 1
                count[changed[end]] -= 1
                start += 1
                end += 1
            elif changed[start] < 2 * changed[end]:
                end += 1
            else:
                start += 1
        # return demo
        return demo if len(demo) == len(changed)/2 else []

File: 2007-find-original-array-from-doubled-array/2007-find-original-array-from-doubled-array.cpp
class Solution {
public:
    vector<int> findOriginalArray(vector<int>& changed) {
        map<int, int> cnt;
        sort(changed.begin(), changed.end());
        vector<int> ans;
        
        for(auto s:changed)
            cnt[s]++;
        
        for (auto s: changed) {
            if (cnt[s]-- > 0) {
                if (cnt[2 * s]-- > 0)
                    ans.push_back(s);
                else
                    return {};
            }
        }
        
        return ans;
    }
};